packages/cardpay-reward-programs/tests/rules/test_staking.py

```diff
@@ -0,0 +1,492 @@
+block_number_so_st = st.integers(min_value=0, max_value=1)
+block_number_st = st.integers(min_value=1, max_value=END_BLOCK + 50)
+balance_st = st.integers(min_value=1, max_value=24032768)
+
+safe_owner_columns = {
+    "owner": {"elements": owner_st, "unique": True},
+    "safe": {"elements": safe_st, "unique": True},
+    "type": {"elements": depot_st},
+    "_block_number": {"elements": block_number_so_st},
+}
+
+token_holder_columns = {
+    "_block_number": {"elements": block_number_st, "unique": True},
+    "safe": {"elements": safe_st},
+    "token": {"elements": card_st},
+    "balance_downscale_e9_uint64": {"elements": balance_st},
+}
+
+safe_owner_df = data_frames(
+    index=range_indexes(min_size=3, max_size=3),
+    columns=[column(key, **value) for key, value in safe_owner_columns.items()],
+)
+
+token_holder_df = data_frames(
+    index=range_indexes(min_size=3, max_size=4),
+    columns=[column(key, **value) for key, value in token_holder_columns.items()],
+)
+
+
+def create_rule(
+    monkeypatch,
+    fake_data_token_holder,
+    fake_data_safe_owner,
+    core_config_overrides={},
+    user_config_overrides={},
+):
+    core_config = {
+        "payment_cycle_length": CYCLE_LENGTH,
+        "start_block": 0,
+        "end_block": END_BLOCK,
+        "subgraph_config_locations": {
+            "prepaid_card_payment": "s3://cardpay-staging-partitioned-graph-data/data/prepaid_card_payments/0.0.3/",
+            "spend_accumulation": "s3://tall-data-dev/paulin/spend_accumulation/0.0.1/",
+            "safe_owner": "s3://tall-data-dev/paulin/safe_owner/0.0.1/",
+        },
+    }
+
+    core_config.update(core_config_overrides)
+    user_config = {"token": "card-0", "duration": 30, "interest_rate_monthly": 0.06}
+
+    user_config.update(user_config_overrides)
+    con = duckdb.connect(database=":memory:", read_only=False)
+    con.execute(
+        f"""create table {token_holder_table} as select * from fake_data_token_holder"""
+    )
+    con.execute(
+        f"""create table {safe_ownership_table} as select * from fake_data_safe_owner"""
+    )
+
+    def table_query(
+        self, config_name, table_name, min_partition: int, max_partition: int
+    ):
+
+        if table_name == "token_holder":
+            return token_holder_table
+        elif table_name == "safe_owner":
+            return safe_ownership_table
+        else:
+            raise NameError
+
+    def run_query(self, table_query, vars, aux_table_query):
+        con.execute(self.sql(token_holder_table, safe_ownership_table), vars)
+        return con.fetchdf()
+
+    monkeypatch.setattr(staking.Staking, "_get_table_query", table_query)
+    monkeypatch.setattr(staking.Staking, "run_query", run_query)
+
+    rule = staking.Staking(core_config, user_config)
+    return rule
+
+
+def get_amount(result, payee, idx):
+    idx = result.index[result["payee"] == payee].tolist()[0]
+    return result.loc[idx, "amount"]
+
+
+def test_correct_calc_rewards(monkeypatch):
```

Comment:
would be good to have a test that the rewards for DATA 
- block 0, balance:100
- block 30, balance: 100

- block 0, balance:100
- block 10, balance: 100
- block 20, balance: 100
- block 30, balance: 100

will give the same rewards in total

Reply:
yeah good idea, will add.